File: python/neuroglancer/credentials_provider.py
# @license
# Copyright 2017 Google Inc.
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import absolute_import

import concurrent.futures
import threading

from .futures import future_then_immediate

class CredentialsManager(object):
    def __init__(self):
        self._providers = dict()

    def register(self, key, credentials_provider_getter):
        self._providers[key] = credentials_provider_getter

    def get(self, key, parameters):
        return self._providers[key](parameters)


class CredentialsProvider(object):
    next_generation = 0
    next_generation_lock = threading.Lock()

    def __init__(self):
        self.credentials = None
        self.future = None
        self._lock = threading.Lock()

    def get(self, invalid_generation=None):
        with self._lock:
            if self.future is not None and (self.credentials is None or
                                            invalid_generation != self.credentials['generation']):
                return self.future
            self.credentials = None

            def attach_generation_and_save_credentials(credentials):
                with self._lock:
                    with CredentialsProvider.next_generation_lock:
                        CredentialsProvider.next_generation += 1
                        next_generation = CredentialsProvider.next_generation
                    credentials_with_generation = dict(
                        credentials=credentials, generation=next_generation)
                    self.credentials = credentials_with_generation
                    return credentials_with_generation

            self.future = future_then_immediate(self.get_new(),
                                                attach_generation_and_save_credentials)
            return self.future

    def get_new(self):
        raise NotImplementedError

File: src/neuroglancer/python_integration/credentials_provider.ts
/**
 * @license
 * Copyright 2017 Google Inc.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * @file Implementation of a CredentialsProvider based on an input and output TrackableValue.
 */

import {AnonymousFirstCredentialsProvider, CredentialsManager, CredentialsProvider, CredentialsWithGeneration, makeCachedCredentialsGetter} from 'neuroglancer/credentials_provider';
import {Client} from 'neuroglancer/python_integration/api';
import {fetchOk} from 'neuroglancer/util/http_request';
import {stableStringify, verifyInt, verifyObject, verifyObjectProperty} from 'neuroglancer/util/json';
import {Memoize} from 'neuroglancer/util/memoize';

class PythonCredentialsProvider<Credentials> extends CredentialsProvider<Credentials> {
  constructor(private client: Client, private key: string, private parameters: any) {
    super();
  }

  get = makeCachedCredentialsGetter(
      async(invalidCredentials?: CredentialsWithGeneration<Credentials>): Promise<
          CredentialsWithGeneration<Credentials>> => {
        const response = await fetchOk(this.client.urls.credentials, {
          method: 'POST',
          body: JSON.stringify(
              {key: this.key, parameters: this.parameters, invalid: invalidCredentials?.generation})
        });
        const json = await response.json();
        verifyObject(json);
        const generation = verifyObjectProperty(json, 'generation', verifyInt);
        const credentials = json['credentials'] as Credentials;
        return {generation, credentials};
      });
}


class GcsCredentialsProvider extends AnonymousFirstCredentialsProvider<any> {
  constructor(baseProvider: CredentialsProvider<any>) {
    super(baseProvider, {accessToken: '', tokenType: ''});
  }
}

export class PythonCredentialsManager implements CredentialsManager {
  constructor(private client: Client) {}
  private memoize = new Memoize<string, CredentialsProvider<any>>();

  getCredentialsProvider<Credentials>(key: string, parameters?: any) {
    if (parameters === undefined) {
      parameters = null;
    }
    const combinedKey = stableStringify({key, parameters});
    return this.memoize.get(combinedKey, () => {
      const provider = new PythonCredentialsProvider<Credentials>(this.client, key, parameters);
      if (key === 'gcs') {
        return new GcsCredentialsProvider(provider);
      }
      return provider;
    });
  }
}
